### PR TITLE
Converted HashInnerJoinBatchIterator to implement the block hash join

### PR DESCRIPTION
**Pull request by andreidan**

### CODE REVIEW

File: dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java [hunk 1]
```diff
@@ -37,12 +38,12 @@
  * <pre>
  *     Build Phase:
  *     for (leftRow in left) {
- *         calculate hash and put in Buffer (HashMap)
+ *         calculate hash and put in Buffer (HashMap) until the blockSize is reached
  *     }
- *     // All rows from left processed and inserted in the Buffer
+ *     // blockSize number of rows from left processed and inserted in the Buffer
```
Comment (seut) [inline dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java]: reads weird and confuses, would drop this sentence at all.

File: sql/src/main/java/io/crate/planner/operators/HashJoin.java [hunk 2]
```diff
@@ -172,7 +173,7 @@ public ExecutionPlan build(PlannerContext plannerContext,
             leftHashInputs,
             rightHashInputs,
             Symbols.typeView(lhs.outputs()),
-            lhs.numExpectedRows());
+            Paging.PAGE_SIZE);
```
Comment (seut) [inline sql/src/main/java/io/crate/planner/operators/HashJoin.java]: what if the page size was updated and adjusted?

File: dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java [hunk 3]
```diff
@@ -115,16 +125,35 @@ public void moveToStart() {
...
+
+    private boolean buildBufferAndMatchRight() {
         if (activeIt == left) {
             while (left.moveNext()) {
                 Object[] currentRow = left.currentElement().materialize();
                 int hash = hashBuilderForLeft.apply(left.currentElement());
                 addToBuffer(currentRow, hash);
             }
-            if (left.allLoaded()) {
+            if (buffer.size() == blockSize || left.allLoaded()) {
```
Comment (matriv) [inline dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java]: I think this is not correct. If left.moveNext() returns false means the page is loaded and we need to switch to the right. The `buffer.size() == blockSize()` should be in the while loop above to exit earlier (as "safety net").

Comment (andreidan) [reply dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java]: I don't think we should operate under the assumption that an iterator batch size is the same as the page size. I think we should load up the buffer (even if it means loading multiple left iterator batches)

File: sql/src/main/java/io/crate/planner/operators/HashJoin.java (hunk 2 above)
Comment (andreidan) [reply sql/src/main/java/io/crate/planner/operators/HashJoin.java]: this just a first step. adjustable page size and HEAP based math are different WPs

**Comment by seut** [comment]
pls check test failure

**Comment by andreidan** [comment]
please retest this

**Comment by andreidan** [comment]
retest this please

**Pull request merged**